pkg/master/master.go

```diff
@@ -779,6 +785,52 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
 	return version
 }
 
+func (m *Master) InstallThirdPartyAPI(rsrc *expapi.ThirdPartyResource) error {
+	kind, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(rsrc)
+	if err != nil {
+		return err
+	}
+	thirdparty := m.thirdpartyapi(group, kind, rsrc.Versions[0].Name)
+	if err := thirdparty.InstallREST(m.handlerContainer); err != nil {
+		glog.Fatalf("Unable to setup thirdparty api: %v", err)
+	}
+	thirdPartyPrefix := "/thirdparty/" + group + "/"
+	apiserver.AddApiWebService(m.handlerContainer, thirdPartyPrefix, []string{rsrc.Versions[0].Name})
+	thirdPartyRequestInfoResolver := &apiserver.APIRequestInfoResolver{APIPrefixes: util.NewStringSet(strings.TrimPrefix(group, "/")), RestMapper: thirdparty.Mapper}
+	apiserver.InstallServiceErrorHandler(m.handlerContainer, thirdPartyRequestInfoResolver, []string{thirdparty.Version})
+	return nil
+}
+
+func (m *Master) thirdpartyapi(group, kind, version string) *apiserver.APIGroupVersion {
+	resourceStorage := thirdpartyresourcedataetcd.NewREST(m.thirdPartyStorage, group, kind)
+
+	apiRoot := "/thirdparty/" + group + "/"
+
+	storage := map[string]rest.Storage{
+		strings.ToLower(kind) + "s": resourceStorage,
+	}
+
+	return &apiserver.APIGroupVersion{
+		Root: apiRoot,
+
+		Creater:   api.Scheme,
+		Convertor: api.Scheme,
+		Typer:     api.Scheme,
+
+		Mapper:  thirdpartyresourcedata.NewMapper(explatest.RESTMapper, kind),
+		Codec:   explatest.Codec,
+		Linker:  explatest.SelfLinker,
+		Storage: storage,
+		Version: version,
```

Comment:
You are right.
Then we need to modify this line: https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/api_installer.go#L132, here kind = ThirdPartyResourceData. If ThirdPartyResource.Versions[0].NAME is different from the versions supported by experimental group, this causes error. See https://github.com/kubernetes/kubernetes/pull/13000/files#r38152090.